docs/proposals/proxy-extensions.md

```diff
@@ -0,0 +1,363 @@
+---
+title: Reverse Proxy Extensions
+
+authors:
+- "@leoluz"
+
+sponsors:
+- TBD
+
+reviewers:
+- TBD
+
+approvers:
+- TBD
+
+creation-date: 2022-07-23
+
+---
+
+# Reverse-Proxy Extensions support for Argo CD
+
+Enable UI extensions to use a backend service.
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+* [Goals](#goals)
+* [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Use cases](#use-cases)
+    * [Security Considerations](#security-considerations)
+    * [Risks and Mitigations](#risks-and-mitigations)
+    * [Upgrade / Downgrade](#upgrade--downgrade)
+* [Drawbacks](#drawbacks)
+* [Open Questions](#open-questions)
+
+---
+
+## Summary
+
+Argo CD currently supports the creation of [UI extensions][1] allowing
+developers to define the visual content of the "more" tab inside
+a specific resource. Developers are able to access the resource state to
+build the UI. However, currently it isn't possible to use a backend
+service to provide additional functionality to extensions. This proposal
+defines a new reverse proxy feature in Argo CD, allowing developers to
+create a backend service that can be used in UI extensions. Extensions
+backend code will live outside Argo CD main repository.
+
+## Motivation
+
+The initiative to implement the anomaly detection capability in Argo CD
+highlighted the need to improve the existing UI extensions feature. The
+new capability will required the UI to have access to data that isn't
+available as part of Application's owned resources. It is necessary to
+access an API defined by the extension's development team so the proper
+information can be displayed.
+
+## Goals
+
+The following goals are desired but not necessarily all must be
+implemented in a given Argo CD release:
+
+#### [G-1] Argo CD (API Server) must have low performance impact when running extensions
+
+Argo CD API server is a critical component as it serves all APIs used by
+the CLI as well as the UI. The Argo CD team has no controll over what is
+going to be executed in extension's backend service. Thus it is important
+that the reverse proxy implementation to cause the lowest possible impact
+in the API server while processing high latency requests.
+
+Possible solutions:
+- Implement a rate limit layer to protect Argo CD API server
+- Implement configurable different types of timeouts (idle connection, duration, etc) between Argo CD API server and backend services.
```

Comment:
@alexmt and @alexec convinced me that it wouldn't be necessary to have a separate Pod just for the proxy. It would increase architectural complexity and less user friendly setup. Having a sane default rate limiting and timeouts configuration will mitigate this issue at a lower cost.

Reply:
Makes sense! Are there any design decisions we can make now to make it _easier_ to transition to a separate Pod if we later feel the need? (Seems like it oughta be fairly easy, but maybe worth thinking about.)

Reply:
I am not sure breaking it in a separate Pod will be a trivial task to do. So I can mention it as a possible solution in this section but I don't think it worth spending a lot of time thinking about it. WDYT?

Reply:
Sounds good!